Replace `privileged_tests` build tag with `PRIVILEGED_TESTS` environment variable

ti-mo opened the pull request:
This is going to ping all SIGs, sorry about that. This PR replaces the `privileged_tests` build tag with test helpers that skip the test or check unless the the `PRIVILEGED_TESTS` environment variable is set.

Using build tags for splitting up a test suite and code in general has a few drawbacks:

- Multiple `go build`s and `go test`s need to be run to ensure all code compiles. This compounds as the matrix of target GOOS and GOARCH values grows, which it has over the past few months/years.
- gopls needs to be configured with -tags=privileged_tests, but now you no longer get code completion in non-privileged tests, which is the majority of files in the codebase.
- Same goes for linters and formatters that are build-tag aware, like golangci-lint and goimports. One needs to pick for which build tags they want to receive warnings and errors.

Why not test flags instead? Here's [an opinion piece](https://peter.bourgon.org/blog/2021/04/02/dont-use-build-tags-for-integration-tests.html) on why environment variables are preferred. The article raises some valid points, but the real case against using test flags is that they need to be declared in every single directory containing a .go file. If there's a package without a flag defined, this is the result:

```
~ go test ./... -run xxxx -privileged
...
flag provided but not defined: -privileged
FAIL	github.com/cilium/cilium/bugtool/cmd	0.006s
```

A three-line test helper is all we really need to accomplish this with an environment variable. Moreover, env vars are easy to add to a build/test context from a Makefile, and they will automatically apply to any and all child processes. Marking privileged tests using a helper gives us the flexibility of changing the underlying mechanism of the helper. This PR proposes and an env var, but we could just as well use another trigger down the line like checking the current UID or checking caps using `capsh`, though that can cause tests to run unintentionally.

```release-note
Replace `privileged_tests` build tag with `PRIVILEGED_TESTS` environment variable
```

ti-mo committed:
*: remove privileged_test build tag in favor of PRIVILEGED_TESTS env

Using build tags for splitting up a test suite and code in general has
a few drawbacks:

- Multiple `go build`s and `go test`s need to be run to ensure all code
  compiles. This compounds as the matrix of target GOOS and GOARCH values
  grows, which it has over the past few months/years.
- gopls needs to be configured with -tags=privileged_tests, but now you
  no longer get code completion in non-privileged tests, which is the
  majority of files in the codebase.
- Same goes for linters and formatters that are build-tag aware, like
  golangci-lint and goimports. One needs to pick for which build tags they
  want to receive warnings and errors.

An opinion piece on why environment variables are preferred:
https://peter.bourgon.org/blog/2021/04/02/dont-use-build-tags-for-integration-tests.html

The article raises some valid points, but the real case against test flags
is because they need to be declared in every single directory containing
a .go file. If there's a package without a flag defined, this is the result:

```
~ go test ./... -run xxxx -privileged
...
flag provided but not defined: -privileged
FAIL	github.com/cilium/cilium/bugtool/cmd	0.006s
```

A three-line test helper is all we really need to accomplish this with an
environment variable. Moreover, env vars are easy to add to a build/test
context from a Makefile, and they will automatically apply to any and all
child processes.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
contrib: invert functionality of check-missing-tags-in-tests.sh

Since the privileged_tests tag has been removed and deprecated,
also remove the script that requires build tags to be present and
replace it with a script that checks the opposite.

This should prevent long-living branches merged at a future point
from reintroducing the tag into test files unnoticed.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
*: replace TestMain() with SetUp/TearDownSuite in privileged tests

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
*: address latent linter warnings after removing build tags

Now the privileged_tests build tags have been removed from these files,
they are once again considered by build tooling. Address the linter warnings
this surfaced.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
Makefile: run privileged tests without -check.vv

Historically, all packages containing privileged tests have made use of gocheck
in one way or another, which takes care of declaring the 'check.vv' flag in
the test binaries of those packages.

However, due to the removal of the priv-test package filtering mechanism in a
prior commit, the test run now also contains packages that don't use gocheck,
resulting in an error due to the -check.vv flag not being known in all
packages.

This patch removes -check.vv from privileged test runs, as using gocheck should
not be a requirement for writing tests.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
pkg/pidfile: remove TestKillPidfilePermissionDenied

Running privileged tests by setting PRIVILEGED_TESTS now no longer disables
unprivileged tests, meaning all tests must be able to run as privileged.

Removing a file with 000 permissions does not generate a 'permission denied'
error when run as root, causing this test to break.Presumably, this test was
added to ensure only errors other than os.ErrNotExist are propagated to the
caller.

The preceding test TestKillPidfileNotExist already runs a positive test on
this behaviour, so TestKillPidfilePermissionDenied is not strictly necessary.

Remove the test so the test suite passes when run with privileges.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
testutils: wait for up to one second in TestCondition

This failed in CI with:

```
condition_test.go:28:
    c.Assert(WaitUntil(countTo5, 100*time.Millisecond), IsNil)
... value *errors.errorString = &errors.errorString{s:
      "timeout reached while waiting for condition"}
      ("timeout reached while waiting for condition")
```

Bump the timeout to 1 second to hopefully always have this complete
without timeouts.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
Makefile: set PRIVILEGED_TESTS env in tests-privileged target

Instead of specifying build tags, privileged tests are now run by setting the
PRIVILEGED_TESTS environment variable added in a prior commit.

Since packages can now no longer be filtered by 'privileged_tests' build
tags, remove PRIV_TEST_PKGS.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
rate: replace flaky test in TestMeanProcessingDuration

Testing the results of calcMeanDuration() is already done in
TestCalcMeanDuration. Depending on the accuracy of sleeps in programs
invites inconsistent behaviour and flakes, especially when running them
in high-load environments like CI VMs. What this really exercises is the
Go runtime and OS schedulers and the underlying virt/hardware.

Shorten the sleeps in the test cycles, check that a mean processing duration
was recorded (whatever its value may be), and add a test case to
TestCalcMeanDuration to compensate somewhat. Crafting more test cases is
tricky due to the API returning floats that are always off by fractions.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
pkg/ipam: relax wait for address expiration timer

Similar to previous commits, this prevents a flake caused by jittery
OS timers in Vagrant VMs during test execution.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
Documentation: document PRIVILEGED_TESTS env variable

Signed-off-by: Timo Beckers <timo@isovalent.com>

joestringer commented on Makefile:
```diff
@@ -125,17 +125,13 @@ build-container: check-sources ## Builds components required for cilium-agent co
-		PATH=$(PATH):$(ROOT_DIR)/bpf $(GO_TEST) $(TEST_LDFLAGS) $$pkg $(GOTEST_UNIT_BASE) $(GOTEST_COVER_OPTS) -coverpkg $$pkg \
+		PATH=$(PATH):$(ROOT_DIR)/bpf \
+		PRIVILEGED_TESTS=true \
+		$(GO_TEST) $(TEST_LDFLAGS) $$pkg $(GOTEST_UNIT_BASE) $(GOTEST_COVER_OPTS) -coverpkg $$pkg \
```
Does this mean that the `tests-privileged` now runs _all_ tests, not just the privileged ones?

Currently we run the unprivileged tests in Travis and run the privileged tests in the ginkgo/jenkins environments, so if this target now includes unprivileged tests then it will likely extend the test runtime by 5 minutes or so, for results we already get from another mechanism. Do you think there's a way we can avoid this with the new mechanism?

joestringer replied:
Couple of possible areas to explore:
- Ensure there is a way to skip unprivileged tests, so that jenkins test doesn't need to run those additional tests that are already run elsewhere
- Move privileged tests out of the jenkins CI into a github action, so we don't extend the jenkins test job runtimes but instead the privileged tests run in parallel with other jobs

ti-mo replied:
@joestringer Thanks for taking a look! I had all the points you raised in mind while working on this.

First,

> Does this mean that the `tests-privileged` now runs all tests, not just the privileged ones?

Yes! And that's fine, because unit ('unprivileged') tests are actually very fast to run, as they should be. The most expensive part of running our test suite is building test binaries. Being able to run all unit, privileged and integration tests in one go is one of the end goals of these changes.

With build cache (4.38s):
```
~ time go test ./...
...
go test ./...  29.60s user 4.38s system 763% cpu 4.450 total
```

Without build cache:
```
~ go clean -testcache ./...
~ time go test ./...
go test ./...  245.62s user 25.07s system 835% cpu 32.383 total
```

This is on a Ryzen 3700X (16 threads) desktop CPU with fast memory, with all source files in OS cache. Even on this machine, build time accounts for 87% of total test suite runtime. I would expect this number to increase even further on cloud VMs with noisy neighbours and slower i/o.

IMO the only proper way of skipping unprivileged tests would be by consistenly naming test functions, e.g. `func Test(Un)PrivilegedFoo()` so they can be filtered out by `go test -run=TestPrivileged`. This is not a realistic goal given the amount of testing frameworks in use throughout the codebase that all have their own quirky way of doing things.

In conclusion: always running unit tests is fine and barely amounts to a rounding error on a total privileged test runtime of 2.5+ hours. It doesn't weigh up against renaming all test functions in existence, nor to adding test helpers to all of them.

---

Second, regarding Jenkins/Travis/GHA, ...

For me, the next step would be cutting out ginkgo from running privileged tests on Jenkins. As far as I can see, it's currently only used for invoking vagrant. The code is inscrutable at best, and heavily interferes in runtime test output, especially when failures occur.

Note that we need nested virt for privileged tests, so we'd need self-hosted GHA runners first.

We'll get there, this is just a first attempt at moving the needle. :slightly_smiling_face:

joestringer replied:
:+1: fair enough. I'm always worried about "just a couple more minutes" on the overall build time explicitly because we mostly just tolerate the test length and we don't grapple with how to make it shorter. So any time a PR like this comes up, the easiest solution is just to tolerate making our dev cycle even longer.

That said, in general we're currently reviewing the jenkins tests for which ones we can convert to other forms of testing / delete so we should be able to make up for this additional test run time pretty soon:tm:. And yeah it's not really fair to hold up this particular PR if this is a general problem with the testing code in Cilium.

> For me, the next step would be cutting out ginkgo from running privileged tests on Jenkins. As far as I can see, it's currently only used for invoking vagrant. The code is inscrutable at best, and heavily interferes in runtime test output, especially when failures occur.

I think the only reason we do it on Jenkins today is to be able to run the tests in a privileged environment. Wherever we can get that, we can run it there.

joestringer reviewed (commented):
There's some nice properties provided by this proposal :+1: 

My main concern is around extending the testsuite runtime by running more tests when we don't need to. Comment below.

ti-mo committed:
daemon: repopulate daemon config before all KPRSuite checks

The read/set mechanism was causing issues due to global state:

```
kube_proxy_replacement_test.go:276:
	     testCase.out.verify(c)
	 kube_proxy_replacement_test.go:73:
	     c.Assert(err, ErrorMatches, cfg.expectedErrorRegex)
	 ... error string = "Invalid value for --node-port-mode: "
	 ... regex string = ""
```

Signed-off-by: Timo Beckers <timo@isovalent.com>

nebril commented on .golangci.yaml:
```diff
@@ -13,8 +13,8 @@ run:
   tests: true
 
   # list of build tags, all linters use it. Default is empty list.
-  build-tags:
```
Should we just remove it?

ti-mo replied:
Done.

nebril reviewed (approved):
:rocket:

ti-mo committed:
Vagrantfiles: enable High-Precision Event Timers in VirtualBox

I've noticed some strange behaviour regarding timer interrupts
when running within VirtualBox VMs. Consider this simple Go program:

```
func main() {
	for {
		start := time.Now()
		time.Sleep(100 * time.Millisecond)
		fmt.Println("Slept", time.Since(start))
	}
}
```

Run under VirtualBox, the output is extremely jittery:
```
vagrant@bullseye:~$ go run main.go
Slept 140.063453ms
Slept 108.029166ms
Slept 159.01884ms
Slept 511.263485ms
...
```

In some outliers, I've noticed a 100ms sleep take up to 3000ms to resume.
Enabling HPET on the VM makes these sleeps take between 100-105ms as expected.

strace shows many more calls to nanosleep() compared to when the program is
run on bare metal or KVM. This may hint at a bad interaction between the Go
scheduler and VirtualBox' default timer infrastructure being consistently
jittery and laggy. May be worth some future investigation, but with KVM
migration efforts underway, we're likely to move off of VBox anyway.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
stream: operators_test: remove TestThrottle OS timer dependency

This patch makes TestThrottle no longer rely on time advancing correctly,
which is not a given while running in a VirtualBox VM.

Assume that the underlying packages and infrastructure is behaving as intended.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>

gandro reviewed (approved):
Looks good for the changes to my codeowner groups. I did not do a review of the overall new mechanism and if it makes sense.
